Modules/signalmodule.c

```diff
@@ -1798,6 +1798,19 @@ int
 PyErr_CheckSignals(void)
 {
     PyThreadState *tstate = _PyThreadState_GET();
+
+    /* Opportunistically check if the GC is scheduled to run and run it
+       if we have a request. This is done here because native code needs
+       to call this API if is going to run for some time without executing
+       Python code to ensure signals are handled. Checking for the GC here
+       allows long running native code to clean cycles created using the C-API
+       even if it doesn't run the evaluation loop */
+    struct _ceval_state *interp_ceval_state = &tstate->interp->ceval;
+    if (_Py_atomic_load_relaxed(&interp_ceval_state->gc_scheduled)) {
+        _Py_atomic_store_relaxed(&interp_ceval_state->gc_scheduled, 0);
```

Comment:
Should we recompute the eval breaker here? Or too difficult?

Reply:
Is a bit annoying because the macros are defined in `ceval.c`. OTOH I don't think is worth it as this is basically independent on the eval breaker mechanism (other than checking a flag that the eval breaker also checks). Not dealing with the eval breaker helps to reason about this